libs/cardano-data/src/Data/Incremental.hs

```diff
@@ -0,0 +1,181 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+
+-- | Introduce the Incremental Lambda Calculus embodied in the ILC class.
+--   Instances for two patterns of use involving Maps.
+module Data.Incremental where
+
+import Control.DeepSeq (NFData (..))
+import Data.Kind
+import Data.Map.Internal (Map (..))
+import Data.Map.Strict
+import qualified Data.Map.Strict as Map
+import GHC.Generics (Generic (..))
+
+-- ===================================================
+-- Incremental lambda calculus
+
+class ILC t where
+  data Diff t :: Type
+  applyDiff :: t -> Diff t -> t
+  extend :: Diff t -> Diff t -> Diff t
+  zero :: Diff t
+
+infixr 0 $$
+($$) :: ILC t => t -> Diff t -> t
+x $$ y = applyDiff x y
+
+-- | Every (Diff t) is a Semigroup
+instance ILC t => Semigroup (Diff t) where
+  x <> y = extend x y
+
+-- | Every (Diff t) is a Monoid
+instance ILC t => Monoid (Diff t) where
+  mempty = zero
+
+-- ========================================
+-- Delta types
+
+-- | Its is deleted, overwritten, or combined using a Monoid
+data MonoidD v = Del | Write !v | Comb !v
+  deriving (Show, Eq, Generic, NFData)
+
+instance (Semigroup t) => Semigroup (MonoidD t) where
```

Comment:
I'm confused about why this is called MonoidD, does it imply it has an identity  element?

Reply:
"monoid of deltas" I suspect

Reply:
so it's a monoid sort of by definition `data MonoidD v = Del | Write !v | Comb !v` ?

Reply:
The intention is that it is the range of a Map that is used to define the Diff of a (Map dom rng)) when the rng is a monoid. So we see that Diff(MonoidMap dom rng) = Map dom (MonoidD (Diff rng)).  Does some thing like MonoidRngD make a better name? One should probably compare  MonoidD to BinaryD, BinaryD has two choices: delete or change. But MonoidD has 3 choices: delete, change, or combine. I will change the names to make this clearer, and add some comments to explain.